src/slicer.cpp

```diff
@@ -51,153 +52,488 @@ void SlicerLayer::makeBasicPolygonLoop(const Mesh* mesh, Polygons& open_polyline
+    connectOpenPolylinesImpl(open_polylines, largest_neglected_gap_second_phase, cell_size, allow_reverse);
+}
+
+void SlicerLayer::stitch(Polygons& open_polylines)
+{
+    bool allow_reverse = true;
+    connectOpenPolylinesImpl(open_polylines, max_stitch1, max_stitch1, allow_reverse);
+}
+
+const SlicerLayer::Terminus SlicerLayer::Terminus::INVALID_TERMINUS{~static_cast<Index>(0U)};
+
+bool SlicerLayer::PossibleStitch::operator<(const PossibleStitch& other) const
+{
+    // better if lower distance
+    if (dist2 > other.dist2)
+    {
+        return true;
+    }
+    else if (dist2 < other.dist2)
     {
-        PolygonRef open_polyline = open_polylines[open_polyline_idx];
-        
-        if (open_polyline.size() < 1) continue;
-        for(unsigned int open_polyline_other_idx = 0; open_polyline_other_idx < open_polylines.size(); open_polyline_other_idx++)
+        return false;
+    }
+
+    // better if in order instead of reversed
+    if (!in_order() && other.in_order())
+    {
+        return true;
+    }
+
+    // better if lower Terminus::Index for terminus_0
+    // This just defines a more total order and isn't strictly necessary.
+    if (terminus_0.asIndex() > other.terminus_0.asIndex())
+    {
+        return true;
+    }
+    else if (terminus_0.asIndex() < other.terminus_0.asIndex())
+    {
+        return false;
+    }
+
+    // better if lower Terminus::Index for terminus_1
+    // This just defines a more total order and isn't strictly necessary.
+    if (terminus_1.asIndex() > other.terminus_1.asIndex())
+    {
+        return true;
+    }
+    else if (terminus_1.asIndex() < other.terminus_1.asIndex())
+    {
+        return false;
+    }
+
+    // The stitches have equal goodness
+    return false;
+}
+
+std::priority_queue<SlicerLayer::PossibleStitch>
+SlicerLayer::findPossibleStitches(
+    const Polygons& open_polylines,
+    coord_t max_dist, coord_t cell_size,
+    bool allow_reverse) const
+{
+    std::priority_queue<PossibleStitch> stitch_queue;
+
+    // maximum distance squared
+    int64_t max_dist2 = max_dist * max_dist;
+
+    // Represents a terminal point of a polyline in open_polylines.
+    struct StitchGridVal
+    {
+        unsigned int polyline_idx;
+        // Depending on the SparseGrid, either the start point or the
+        // end point of the polyline
+        Point polyline_term_pt;
+    };
+
+    struct StitchGridValLocator
+    {
+        Point operator()(const StitchGridVal& val) const
         {
-            PolygonRef open_polyline_other = open_polylines[open_polyline_other_idx];
-            
-            if (open_polyline_other.size() < 1) continue;
-            
-            Point diff = open_polyline.back() - open_polyline_other[0];
+            return val.polyline_term_pt;
+        }
+    };
+
+    // Used to find nearby end points within a fixed maximum radius
+    SparseGridInvasive<StitchGridVal,StitchGridValLocator> grid_ends(cell_size);
```

Comment:
Shouldn't this now just be a `SparseGrid` rather than a `SparseGridInvasive`?
You've now changed the element type to include the point, which is precisely what SparseGrid does.